software/controller/lib/hal/spi.h

```diff
@@ -15,51 +15,76 @@
 
 #pragma once
 
-/// \TODO create more of an abstraction than just providing registers
-
-#include <cstdint>
-
-// [RM] 40.6 SPI Registers (pg 1330)
-struct SpiStruct {
-  struct {
-    uint32_t clock_phase : 1;
-    uint32_t clock_polarity : 1;
-    uint32_t master : 1;
-    uint32_t bitrate : 3;
-    uint32_t enable : 1;
-    uint32_t lsb_first : 1;
-    uint32_t internal_slave_select : 1;
-    uint32_t sw_slave_management : 1;
-    uint32_t rx_only : 1;
-    uint32_t crc_length : 1;
-    uint32_t crc_next : 1;
-    uint32_t crc_enable : 1;
-    uint32_t bidir_output : 1;
-    uint32_t bidir_mode : 1;
-    uint32_t reserved : 16;
-  } control_reg1;  // Control register 1 (SPIx_CR1) [RM] 40.6.1 (pg 1330)
-  struct {
-    uint32_t rx_dma : 1;
-    uint32_t tx_dma : 1;
-    uint32_t ss_output : 1;
-    uint32_t nss_pulse : 1;
-    uint32_t frame_format : 1;
-    uint32_t error_interrupt : 1;
-    uint32_t rx_not_empty_interrupt : 1;
-    uint32_t tx_empty_interrupt : 1;
-    uint32_t data_size : 4;
-    uint32_t fifo_rx_threshold : 1;
-    uint32_t last_dma_rx : 1;
-    uint32_t last_dma_tx : 1;
-    uint32_t reserved : 17;
-  } control2;  // Control register 2 (SPIx_CR2) [RM] 40.6.2
-  uint32_t status;
-  uint32_t data;
-  uint32_t crc_polynomial;
-  uint32_t rx_crc;
-  uint32_t tx_crc;
+// The STM32 chip comes with 3 SPI channels, which can be used to exchange
+// data with peripherals such as our stepper motors.
+//
+// Current implementation does not provide any bufferization, which is done
+// by the user.
+// Also, this only works with DMA, if DMA::Base provided in constructor cannot
+// be mapped to the desired SPI::Base, the transfers will not work.
+
+#include <optional>
+
+#include "dma.h"
+#include "gpio_stm32.h"
+#include "serial_listeners.h"
+
+namespace SPI {
+
+enum class Base { SPI1, SPI2, SPI3 };
+
+// possible bitrates on SPI bus (see [RM] p1131).
+enum class Bitrate : uint8_t {
+  CpuFreqByTwo = 0,
+  CpuFreqByFour = 1,
+  CpuFreqByEight = 2,
+  CpuFreqBySixteen = 3,
+  CpuFreqByThirtyTwo = 4,
+  CpuFreqBySixtyFour = 5,
+  CpuFreqByOneTwentyEight = 6,
+  CpuFreqByTwoFiftySix = 7,
 };
-typedef volatile SpiStruct SpiReg;
-inline SpiReg *const Spi1Base = reinterpret_cast<SpiReg *>(0x40013000);
-inline SpiReg *const Spi2Base = reinterpret_cast<SpiReg *>(0x40003800);
-inline SpiReg *const Spi3Base = reinterpret_cast<SpiReg *>(0x40003C00);
+
+class Channel {
+ public:
+  Channel(Base spi, DMA::Base dma);
+
+  void Initialize(GPIO::Port clock_port, uint8_t clock_pin, GPIO::Port miso_port, uint8_t miso_pin,
```

Comment:
Holy hell this is getting chunky. I see under `system_constants.h` a lot of `port` + `pin`( + `channel`) tuples. Maybe it's worth creating an abstraction to ensapsulate such addresses?

Reply:
Yes, I was thinking the same thing, though I suggest making this a task for later.

Reply:
Sure, we can do this later.